src/components/ProjectList.tsx

```diff
@@ -101,64 +106,97 @@ const ProjectList: React.FC = () => {
-            }
-            return false;
-          });
-
-          const paratimeMatch: boolean =
-            selectedParatimes.length > 0 &&
-            selectedParatimes.some(paratime =>
-              project.paratimes?.includes(paratime)
-            );
-
-
-    return searchMatch && tagsMatch && langsMatch  && paratimeMatch && maintainedByOasisMatch && licenseMatch && sourcesMatch;
+      selectedSources.length === 0 ||
+      selectedSources.some((source) => {
+        if (
+          (source === 'Demo' && project.demoUrl) ||
+          (source === 'Code' && project.codeUrl)
+        ) {
+          return true;
+        }
+        if (
+          source === 'Tutorial' &&
+          Array.isArray(project.tutorials) &&
+          project.tutorials.length > 0
+        ) {
+          return true;
+        }
+        return false;
+      });
+
+    const paratimeMatch: boolean =
+      selectedParatimes.length === 0 ||
+      selectedParatimes.some(paratime =>
+        project.paratimes?.includes(paratime)
+      );
+
+    return searchMatch && tagsMatch && langsMatch && paratimeMatch && maintainedByOasisMatch && licenseMatch && sourcesMatch;
   });
 
+  const tagCounts = tags.reduce((acc, tag) => {
+    acc[tag] = filteredProjects.filter((project) => project.tags.includes(tag)).length;
+    return acc;
+  }, {} as Record<string, number>);
+
+  const langCounts = langs.reduce((acc, lang) => {
+    acc[lang] = filteredProjects.filter((project) => project.languages.includes(lang)).length;
+    return acc;
+  }, {} as Record<string, number>);
+
+  const licenseCounts = licenses.reduce((acc, license) => {
+    acc[license] = filteredProjects.filter((project) => project.license === license).length;
+    return acc;
+  }, {} as Record<string, number>);
+
+  const sourceCounts = ['Demo', 'Code', 'Tutorial'].reduce((acc, source) => {
+    acc[source] = filteredProjects.filter((project) => {
+      if (source === 'Demo') return !!project.demoUrl;
+      if (source === 'Code') return !!project.codeUrl;
+      if (source === 'Tutorial') return Array.isArray(project.tutorials) && project.tutorials.length > 0;
+      return false;
+    }).length;
+    return acc;
+  }, {} as Record<string, number>);
+
+  const paratimeCounts = ['sapphire', 'emerald', 'cipher'].reduce((acc, paratime) => {
+    acc[paratime] = filteredProjects.filter((project) => project.paratimes?.includes(paratime)).length;
+    return acc;
+  }, {} as Record<string, number>);
+  
+  const maintainedByOasisCount = filteredProjects.filter(
+    (project) => project.maintainedByOasis
+  ).length;
 
-  const handleTagClick = (tag: string) => {
-    if (selectedTags.includes(tag)) {
-      setSelectedTags(selectedTags.filter((t) => t !== tag));
-    } else {
-      setSelectedTags([...selectedTags, tag]);
-    }
+  const handleTagClick = (tags: string[]) => {
+    setSelectedTags(tags);
+  };
+
+  const handleSingleTagClick = (tag: string) => {
+    handleTagClick(
```

Comment:
This is nicely simplified! But the function names are confusing, especially `handleTagClick={handleSingleTagClick}`. I would rename:
```diff
- const handleTagClick = (
+ setSelectedTags

- const handleSingleTagClick = (tag: string) => {
-   handleTagClick(
+ const handleTagClick = (tag: string) => {
+   setSelectedTags(

- handleTagClick={handleSingleTagClick}
+ handleTagClick={handleTagClick}

- handleTagClick={handleTagClick}
+ setSelectedTags={setSelectedTags}
```

Reply:
Renamed.